### PR TITLE
feat(ui): UI for Event-source

### LINKED ISSUE #4487: Event source GUI editor
The ability to create and edit event sources in the U.

### PR DESCRIPTION
**Pull request by sarabala1979**
Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

closes #4487

### CODE REVIEW

File: pkg/apiclient/eventsource/eventsource.swagger.json [hunk 1]
```diff
@@ -0,0 +1,4555 @@
+{
```
Comment (alexec) [inline pkg/apiclient/eventsource/eventsource.swagger.json:1]: add this to gitignore and delete it

File: ui/src/app/event-sources/components/event-source-details/event-source-details.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                                    .then(() => setEdited(false))
+                                    .then(() => setError(null))
+                                    .catch(setError)
+                        },
+                        {
+                            title: 'Delete',
+                            iconClassName: 'fa fa-trash',
+                            disabled: edited,
+                            action: () => {
+                                if (!confirm('Are you sure you want to delete this event source?\nThere is no undo.')) {
```
Comment (alexec) [inline ui/src/app/event-sources/components/event-source-details/event-source-details.tsx:80]: minor - should be trying to use the `popup.confirm`

File: ui/src/app/event-sources/components/event-source-list/event-source-list.tsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+import {Context} from '../../../shared/context';
+import {historyUrl} from '../../../shared/history';
+import {services} from '../../../shared/services';
+import {EventsPanel} from '../../../workflows/components/events-panel';
+import {EventSourceCreator} from '../event-source-creator';
+import {EventSourceLogsViewer} from '../event-source-log-viewer';
+
+const learnMore = <a href='https://argoproj.github.io/argo-events/concepts/eventsource/'>Learn more</a>;
+
+// @ts-ignore
```
Comment (alexec) [inline ui/src/app/event-sources/components/event-source-list/event-source-list.tsx:24]: should not use `@ts-ignore`

File: ui/src/app/event-sources/components/event-source-summary-panel.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import * as React from 'react';
+
+import {ExampleManifests} from '../../shared/components/example-manifests';
+import {ResourceEditor} from '../../shared/components/resource-editor/resource-editor';
+import {Timestamp} from '../../shared/components/timestamp';
+import {services} from '../../shared/services';
+// @ts-ignore
```
Comment (alexec) [inline ui/src/app/event-sources/components/event-source-summary-panel.tsx:7]: ts-ignore should be removed and errors fixed

File: ui/src/app/event-sources/components/event-source-summary-panel.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                    <ResourceEditor
+                        kind='Eventsource'
+                        title='Update Event source'
+                        value={props.eventSource}
+                        onSubmit={(value: EventSource ) =>
+                            services.eventSource.update(value, props.eventSource.metadata.namespace).then(eventSource => props.onChange(eventSource))
+                        }
+                    />
+                    <p>
+                        <ExampleManifests />
```
Comment (alexec) [inline ui/src/app/event-sources/components/event-source-summary-panel.tsx:44]: I'm not sure this applies to event sources, are there examples else where we could point users too?

File: ui/src/app/shared/components/status-notice.tsx [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import * as React from 'react';
```
Comment (alexec) [inline ui/src/app/shared/components/status-notice.tsx:1]: delete file

File: ui/src/app/shared/components/workflow-spec-panel/types.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+export const types: {[type: string]: boolean} = {
```
Comment (alexec) [inline ui/src/app/shared/components/workflow-spec-panel/types.ts:1]: delete file

File: hack/install-argo-events.sh [hunk 8]
```diff
@@ -0,0 +1,43 @@
+set -eu
```
Comment (alexec) [inline hack/install-argo-events.sh:1]: delete file

File: ui/src/app/shared/examples.ts [hunk 9]
```diff
@@ -95,13 +95,26 @@ export const exampleCronWorkflow = (namespace: string): CronWorkflow => ({
...
+    }
+  });
+
 export const exampleSensor = (namespace: string): Sensor => ({
     metadata: {
         name: randomSillyName(),
         namespace,
         labels
     },
-    spec: {
```
Comment (alexec) [inline ui/src/app/shared/examples.ts:104]: ??

File: ui/src/app/shared/services/event-source-service.ts [hunk 10]
```diff
@@ -1,11 +1,37 @@
-import {EventSourceList, EventSourceWatchEvent, LogEntry} from '../../../models/event-source';
+import {EventSource, EventSourceList, EventSourceWatchEvent, LogEntry} from '../../../models/event-source';
 import requests from './requests';
 
 export class EventSourceService {
+    public create(eventSource: EventSource, namespace: string) {
+        const str = JSON.stringify(eventSource, null, 4);
+        // tslint:disable-next-line:no-console
```
Comment (alexec) [inline ui/src/app/shared/services/event-source-service.ts:7]: delete line - fix lint

**Comment by sarabala1979** [comment]
@alexec can you take another look?

File: ui/src/models/index.ts [hunk 11]
```diff
@@ -4,5 +4,7 @@ export * from './workflows';
 export * from './workflow-templates';
 export * from './cron-workflows';
 export * from './cluster-workflow-templates';
+// @ts-ignore
```
Comment (alexec) [inline ui/src/models/index.ts:7]: remove all ts-ignore

**Review by alexec — COMMENTED** [review]
can you please attach screenshots

File: server/eventsource/event_source_server.go [hunk 12]
```diff
@@ -18,6 +18,50 @@ import (
 
 type eventSourceServer struct{}
 
+func (e *eventSourceServer) CreateEventSource(ctx context.Context, in *eventsourcepkg.CreateEventSourceRequest) (*esv1.EventSource, error) {
+	client := auth.GetEventSourceClient(ctx)
```
Comment (alexec) [inline server/eventsource/event_source_server.go:22]: lets get rid of GetOptions and CreateOptions completely - we don't need them

**Comment by sarabala1979** [comment]
image

image

image

image

image

File: ui/src/app/event-sources/components/event-source-details/event-source-details.tsx [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+                                    .then(() => setEdited(false))
+                                    .then(() => setError(null))
+                                    .catch(setError)
+                        },
+                        {
+                            title: 'Delete',
+                            iconClassName: 'fa fa-trash',
+                            disabled: edited,
+                            action: () => {
+                                popup.confirm('Confirm', 'Are you sure you want to delete this event source?\nThere is no undo.').then(yes => {
```
Comment (alexec) [inline ui/src/app/event-sources/components/event-source-details/event-source-details.tsx:100]: this looks different to other uses - making me thing those other uses are wrong - can you please check them and fix if needed?

File: ui/src/app/event-sources/components/event-source-details/event-source-details.tsx [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+                                            .then(() => navigation.goto(uiUrl('event-sources/' + namespace)))
+                                            .then(() => setError(null))
+                                            .catch(setError);
+                                    }
+                                });
+                            }
+                        },
+                        {
+                            title: 'Logs',
+                            iconClassName: 'fa fa-file-alt',
```
Comment (alexec) [inline ui/src/app/event-sources/components/event-source-details/event-source-details.tsx:113]: `fa-bars`

File: ui/src/app/event-sources/components/event-source-details/event-source-details.tsx [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+                    <Loading />
+                ) : (
+                    <EventSourceEditor eventSource={eventSource} onChange={setEventSource} onError={setError} onTabSelected={setTab} selectedTabKey={tab} />
+                )}
+            </>
+            <SlidingPanel isShown={!!selected} onClose={() => setSelectedNode(null)}>
+                {!!selectedNode && (
+                    <div>
+                        <h4>
+                            EventSource/{selected.name}
```
Comment (alexec) [inline ui/src/app/event-sources/components/event-source-details/event-source-details.tsx:134]: I think you can remove `EventSource/` from this line - it is implicit

File: ui/src/app/event-sources/components/event-source-list/event-source-list.tsx [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+                },
+                tools: [<NamespaceFilter key='namespace-filter' value={namespace} onChange={setNamespace} />]
+            }}>
+            <ErrorNotice error={error} />
+            {!eventSources ? (
+                <Loading />
+            ) : eventSources.length === 0 ? (
+                <ZeroState title='No eventsources'>
+                    <p>You can create new EventSource here.</p>
+                    <p>{learnMore}.</p>
```
Comment (alexec) [inline ui/src/app/event-sources/components/event-source-list/event-source-list.tsx:95]: nice

File: ui/src/app/event-sources/components/event-source-log-viewer.tsx [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+        const source = services.eventSource
+            .eventSourcesLogs(namespace, eventSource.metadata.name, parts[0], parts[1], '', 50)
+            .filter(e => !!e)
+            .map(
+                e =>
+                    Object.entries(e)
+                        .map(([key, value]) => key + '=' + value)
+                        .join(', ') + '\n'
+            )
+            .publishReplay()
```
Comment (alexec) [inline ui/src/app/event-sources/components/event-source-log-viewer.tsx:43]: I don't understand this code - can you test it works ok?

File: ui/src/app/event-sources/components/event-source-log-viewer.tsx [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+                                            <a>
+                                                <span title={name}>&nbsp;&nbsp;{name}</span>
+                                            </a>
+                                        </div>
+                                    ))}
+                                </div>
+                            ))}
+                    </div>
+                </div>
+                <div className='columns small-9 medium-10' style={{height: 600}}>
```
Comment (alexec) [inline ui/src/app/event-sources/components/event-source-log-viewer.tsx:90]: remove the 600 here - I've fixed this on origin today

**Review by alexec — APPROVED** [review]
minor comments

File: ui/src/app/event-sources/components/event-source-details/event-source-details.tsx (hunk 13 above)
Comment (sarabala1979) [reply ui/src/app/event-sources/components/event-source-details/event-source-details.tsx:100]: only the workflow page is using this. I will do it separate PR

**Pull request merged**